Issue 245 better logging

patilvishal0597 opened the pull request:
Added a logger util for configuration of logger. Tested logging of invoke, ainvoke, and converse request/responses


### Invoke call:

```
llm = ChatBedrock(
        model_id="us.anthropic.claude-3-5-sonnet-20241022-v2:0",
        region_name="us-east-1",
        model_kwargs={
            "max_tokens": 100,
            "top_p": 0.9,
            "temperature": 0.1,
        },
    )

# input to llm
messages = [
        (
            "system",
            "You are a helpful assistant that translates English to French. Translate the user sentence.",
        ),
        ("human", "I love going out for a walk when the weather is bright and sunny."),
    ]

# Invoke the llm
response = llm.invoke(messages)
```

### Logging output:

 
#### Debug logs with LANGCHAIN_AWS_DEBUG and LANGCHAIN_AWS_DEBUG_ROOT env vars as True
```
...
...
....
2025-01-15 14:30:31 DEBUG | [connectionpool.py:1051] | urllib3.connectionpool - Starting new HTTPS connection (1): bedrock-runtime.us-east-1.amazonaws.com:443
2025-01-15 14:30:33 DEBUG | [connectionpool.py:546] | urllib3.connectionpool - https://bedrock-runtime.us-east-1.amazonaws.com:443 "POST /model/us.anthropic.claude-3-5-sonnet-20241022-v2%3A0/invoke HTTP/11" 200 299
2025-01-15 14:30:33 DEBUG | [parsers.py:241] | botocore.parsers - Response headers: {'Date': 'Wed, 15 Jan 2025 22:30:33 GMT', 'Content-Type': 'application/json', 'Content-Length': '299', 'Connection': 'keep-alive', 'x-amzn-RequestId': 'xxxxxx', 'X-Amzn-Bedrock-Invocation-Latency': '1077', 'X-Amzn-Bedrock-Output-Token-Count': '22', 'X-Amzn-Bedrock-Input-Token-Count': '40'}
2025-01-15 14:30:33 DEBUG | [parsers.py:242] | botocore.parsers - Response body:
<botocore.response.StreamingBody object at 0x108e595d0>
2025-01-15 14:30:33 DEBUG | [hooks.py:238] | botocore.hooks - Event needs-retry.bedrock-runtime.InvokeModel: calling handler <botocore.retryhandler.RetryHandler object at 0x108e58c80>
2025-01-15 14:30:33 DEBUG | [retryhandler.py:211] | botocore.retryhandler - No retry needed.
2025-01-15 14:30:33 INFO | [bedrock.py:603] | root - The output message sent by user: content="J'aime me promener quand il fait beau et ensoleillé." additional_kwargs={'usage': {'prompt_tokens': 40, 'completion_tokens': 22, 'total_tokens': 62}, 'stop_reason': 'end_turn', 'model_id': 'us.anthropic.claude-3-5-sonnet-20241022-v2:0'} response_metadata={} usage_metadata={'input_tokens': 40, 'output_tokens': 22, 'total_tokens': 62}
J'aime me promener quand il fait beau et ensoleillé.
(base) vishankp@7cf34de71c79 aws % 
```

#### Logging output when LANGCHAIN_AWS_DEBUG and LANGCHAIN_AWS_DEBUG_ROOT flags are False and logger is initialized with a module_name:

```
2025-01-15 14:39:40 INFO | langchain_aws.chat_models.bedrock - The input message sent by user: [SystemMessage(content='You are a helpful assistant that translates English to French. Translate the user sentence.', additional_kwargs={}, response_metadata={}), HumanMessage(content='I love going out for a walk when the weather is bright and sunny.', additional_kwargs={}, response_metadata={})]
2025-01-15 14:39:40 ERROR | langchain_aws.chat_models.bedrock - Testing error log
2025-01-15 14:39:41 INFO | langchain_aws.chat_models.bedrock - The output message sent by user: content="J'aime me promener quand il fait beau et ensoleillé." additional_kwargs={'usage': {'prompt_tokens': 40, 'completion_tokens': 22, 'total_tokens': 62}, 'stop_reason': 'end_turn', 'model_id': 'us.anthropic.claude-3-5-sonnet-20241022-v2:0'} response_metadata={} usage_metadata={'input_tokens': 40, 'output_tokens': 22, 'total_tokens': 62}
J'aime me promener quand il fait beau et ensoleillé.
```

guilt commented on libs/aws/langchain_aws/chat_models/bedrock_converse.py:
```diff
@@ -495,13 +498,15 @@ def _generate(
     ) -> ChatResult:
         """Top Level call"""
         bedrock_messages, system = _messages_to_bedrock(messages)
+        # logger.info(f"The input message sent by user: {bedrock_messages}")
```
Please do not add commented code. If you want for it to be called only for fewer times, use `.debug()`

guilt commented on libs/aws/langchain_aws/logger_util.py:
```diff
@@ -0,0 +1,27 @@
+    '''
+    # define handler and formatter
+    handler = logging.StreamHandler()
+    formatter = logging.Formatter("%(asctime)s - %(levelname)s - %(name)s - %(message)s")
```
Suggestion, move these as constants outside. You could let the user pass these formatters if you want as an arg.

```python3
def configure_logger( ... , handler = LANGCHAIN_HANDLER, formatter = LANGCHAIN_FORMATTER, ...)
```

What does the rest of the `langchain` code use for logging?

patilvishal0597 replied:
I don't see any logging in the rest of the langchain aws repo. As per @3coins in [this comment](https://github.com/langchain-ai/langchain-aws/issues/245#issuecomment-2427524855)

Commit:
Addresses logging Issue

Added `logger_util` to enable package and class wide logging in `langchain-aws`.
Added logging for `invoke` and `ainvoke`.

https://github.com/langchain-ai/langchain-aws/issues/245

guilt commented on libs/aws/langchain_aws/logger_util.py:
```diff
@@ -0,0 +1,107 @@
+
+    # If log handler is None, set default stream as std.out
+    if not log_handler:
+        log_handler = logging.StreamHandler()
```
ToDo: What if `log_formatter` is unset?

patilvishal0597 replied:
log_formatter has a DEFAULT_LOG_FORMATTER value that is being set at line 45

guilt replied:
What if user passes `None`?

michaelnchin committed:
Merge branch 'main' into issue-245-better-logging

Commit:
Merge remote-tracking branch 'origin/main' into issue-245-better-logging

Commit:
Simplified logging logic

Commit:
Merge remote-tracking branch 'origin/main' into issue-245-better-logging

Commit:
Linting/formatting changes

guilt commented:
This was already merged on the main branch. 👍🏼
